iox_catalog/src/postgres.rs

```diff
@@ -240,6 +217,143 @@ impl Catalog for PostgresCatalog {
+    info!(application_name=%app_name2, "connected to config store");
+
+    Ok(pool)
+}
+
+/// Creates a new HotSwapPool
+///
+/// This function understands the IDPE specific `dsn-file://` dsn uri scheme
+/// and hot swaps the pool with a new sqlx::Pool when the file changes.
+/// This is useful because the credentials can be rotated by infrastructure
+/// agents while the service is running.
+///
+/// The file is polled for changes every `polling_interval`.
+///
+/// The pool is replaced only once the new pool is successfully created.
+/// The [`new_raw_pool`] function will return a new pool only if the connection
+/// is successfull (see [`sqlx::pool::PoolOptions::test_before_acquire`]).
+async fn new_pool(
+    app_name: &str,
+    schema_name: &str,
+    dsn: &str,
+    polling_interval: Duration,
+) -> Result<HotSwapPool<Postgres>, sqlx::Error> {
+    let app_name: Arc<str> = Arc::from(app_name);
+    let schema_name: Arc<str> = Arc::from(schema_name);
+    let parsed_dsn = match parse_dsn_file(dsn) {
+        Some(filename) => std::fs::read_to_string(&filename)?,
+        None => dsn.to_owned(),
+    };
+    let pool = HotSwapPool::new(new_raw_pool(&app_name, &schema_name, &parsed_dsn).await?);
+
+    if let Some(dsn_file) = parse_dsn_file(dsn) {
+        let pool = pool.clone();
+
+        // TODO(mkm): return a guard that stops this background worker.
+        // We create only one pool per process, but it would be cleaner to be
+        // able to properly destroy the pool. If we don't kill this worker we
+        // effectively keep the pool alive (since it holds a reference to the
+        // Pool) and we also potentially pollute the logs with spurious warnings
+        // if the dsn file disappears (this may be annoying if they show up in the test
+        // logs).
+        tokio::spawn(async move {
+            let mut current_dsn = parsed_dsn.clone();
+            loop {
+                tokio::time::sleep(polling_interval).await;
+
+                async fn try_update(
+                    app_name: &str,
+                    schema_name: &str,
+                    current_dsn: &str,
+                    dsn_file: &str,
+                    pool: &HotSwapPool<Postgres>,
+                ) -> Result<Option<String>, sqlx::Error> {
+                    let new_dsn = std::fs::read_to_string(&dsn_file)?;
+                    if new_dsn == current_dsn {
+                        Ok(None)
+                    } else {
+                        let new_pool = new_raw_pool(app_name, schema_name, &new_dsn).await?;
+                        pool.replace(new_pool);
+                        Ok(Some(new_dsn))
+                    }
+                }
+
+                match try_update(&app_name, &schema_name, &current_dsn, &dsn_file, &pool).await {
+                    Ok(None) => {}
+                    Ok(Some(new_dsn)) => {
+                        info!("replacing hotswap pool");
```

Comment:
```suggestion
                        info!("replaced hotswap pool");
```

This is logged _after_ the call to `pool.replace()` - I'm sure this will confuse someone on day! Might be worth a log before too in case we're picking through logs?